include/at_cmd.h

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#ifndef AT_CMD_INTERFACE_H_
+#define AT_CMD_INTERFACE_H_
+
+/**
+ * @file at_cmd_interface.h
+ *
+ * @defgroup at_cmd_interface AT Command interface driver
+ *
+ * @{
+ *
+ * @brief Public APIs for the AT command interface driver.
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <zephyr/types.h>
+
+/**
+ * @brief AT command return codes
+ */
+enum at_cmd_rc {
+	AT_CMD_OK,
+	AT_CMD_ERROR,
+	AT_CMD_CMS,
+	AT_CMD_CME,
+	AT_CMD_NOTIFICATION,
+};
+
+/**
+ * @typedefs at_cmd_handler_t
+ *
+ * Because this driver let multiple threads share the same socket, it must make
+ * sure that the correct thread gets the correct data returned from the AT
+ * interface. The_cmd_write() function let the user specify the handler that
+ * will process the data from a specific AT command call. Notifications will be
+ * handled by the global handler defined using the at_cmd_set_handler()
+ * function. Both handlers are of the type at_cmd_handler_t.
+ *
+ * @param response     Null terminated string containing the modem message
+ *
+ */
+typedef void (*at_cmd_handler_t)(char *response);
```

Comment:
I would really like to have another parameter in this callback to store user data. Could you please add a `void * p_user_data` parameter, store it and return it in the callback?
This is a common practice for all callbacks. By doing this, the caller can pass data and retrieve it later (like a struct or a class pointer), needed to dispatch the result correctly.

Reply:
Some questions, 

* Where should I send in the p_user_data pointer, at_cmd_write("AT+CFUN=1", p_handler, p_user_data) and at_cmd_set_notification_handler(p_handler, p_user_data)?
* I should probably store separate user_data pointers for notifications and writes?
* What would be typical use-cases for this? I can see one use-case, in the link controller driver where I can send in the "link" variable in the w_lte_lc_init_and_connect function instead of using a static file-scope variable.

Reply:
Can we close this PR today? Can we put this proposal in the backlog and merge the code ass is?